From 8f7c74c760fa934cef9c0aec002550f78d71ac0e Mon Sep 17 00:00:00 2001
From: liangbaihui <baihui.liang@spacemit.com>
Date: Tue, 17 Dec 2024 15:37:38 +0800
Subject: [PATCH 1372/1451] gpu: fix sleep resume failures

add checks for psPowerDevice when the GPU is not fully initialized to avoid null pointer issues.

Change-Id: I7adb7c841d41f82f498e71db1bf2fd37b1049e95
---
 drivers/gpu/drm/img-rogue/power.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/img-rogue/power.c b/drivers/gpu/drm/img-rogue/power.c
index dfd8abac72a3..0f3132ced810 100644
--- a/drivers/gpu/drm/img-rogue/power.c
+++ b/drivers/gpu/drm/img-rogue/power.c
@@ -993,8 +993,14 @@ PVRSRV_ERROR PVRSRVSetDeviceSystemPowerState(PPVRSRV_DEVICE_NODE psDeviceNode,
 	_PVRSRVForcedPowerLock(psDeviceNode);
 	psPowerDevice = psDeviceNode->psPowerDev;
 
-	eNewDevicePowerState = _IsSystemStatePowered(eNewSysPowerState)
-	    ? psPowerDevice->eDefaultPowerState : PVRSRV_DEV_POWER_STATE_OFF;
+	if (psPowerDevice){
+		eNewDevicePowerState = _IsSystemStatePowered(eNewSysPowerState)
+			? psPowerDevice->eDefaultPowerState : PVRSRV_DEV_POWER_STATE_OFF;
+	}
+	else{
+		eNewDevicePowerState = _IsSystemStatePowered(eNewSysPowerState)
+			? PVRSRV_DEV_POWER_STATE_DEFAULT : PVRSRV_DEV_POWER_STATE_OFF;
+	}
 
 	/* If setting devices to default state, force idle all devices whose default state is off */
 	pfnIsDefaultStateOff =
@@ -1030,8 +1036,8 @@ PVRSRV_ERROR PVRSRVSetDeviceSystemPowerState(PPVRSRV_DEVICE_NODE psDeviceNode,
 	}
 
 	/* Call power function if the state change or if this is an OS request. */
-	if (OSAtomicRead(&psPowerDevice->eCurrentPowerState) != eNewDevicePowerState ||
-	    BITMASK_ANY(ePwrFlags, PVRSRV_POWER_FLAGS_OSPM_SUSPEND_REQ | PVRSRV_POWER_FLAGS_OSPM_RESUME_REQ))
+	if (psPowerDevice && (OSAtomicRead(&psPowerDevice->eCurrentPowerState) != eNewDevicePowerState ||
+	    BITMASK_ANY(ePwrFlags, PVRSRV_POWER_FLAGS_OSPM_SUSPEND_REQ | PVRSRV_POWER_FLAGS_OSPM_RESUME_REQ)))
 	{
 		eError = PVRSRVDeviceSystemPrePowerStateKM(psPowerDevice,
 												   eNewDevicePowerState,
-- 
2.47.0

